generators/add-licence-headers/index.js

```diff
@@ -0,0 +1,202 @@
+    }).catch((err) => this.log(err));
+  }
+
+  writing() {
+    this._generateComments();
+
+    try {
+      const sourceFolders = this._getSourceFolders();
+
+      sourceFolders.forEach((folderName) => {
+        const pluginPath = this.destinationPath(folderName);
+        if (fs.existsSync(pluginPath)) {
+          this.fileTypes.forEach((type) => this._addComments(pluginPath, type));
+        }
+      });
+    } catch (e) {
+      this._abort(e);
+    }
+  }
+
+  _readLicenceFile() {
+    try {
+      this.licenceText = this.fs.read(this.licencePath);
+    } catch (e) {
+      return this._abort(e);
+    }
+  }
+
+  _generateComments() {
+    // get maximum line length by looping through all lines, returning the line length and passing them to Math.max
+    const lineArray = this.licenceText.split('\n');
+    const maxLineLength = Math.max(...lineArray.map((line) => line.length));
+
+    const align = (string, threshold, timesSpaces) => {
+      if (!timesSpaces) {
+        timesSpaces = threshold;
+      }
+
+      if (string.length > threshold) {
+        return ' '.repeat(timesSpaces);
+      }
+      return '';
+    };
+
+    Object.keys(comments).forEach((fileType) => {
+      const commentStyle = comments[fileType];
+      let comment = commentStyle.begin + commentStyle.body.repeat(maxLineLength) + '\n';
+      lineArray.forEach((line) => {
+        // only add a space character when the opening comment contains more than one charcters (e.g. to align the asterisks in .ts files vs. aligning the hashes in .py files)
+        comment += align(commentStyle.begin, commentStyle.aligningSpaces);
+        comment += `${commentStyle.body}${align(line, 0, 1) + line}\n`;
+      });
+
+      comment += align(commentStyle.begin, commentStyle.aligningSpaces);
+      comment += `${commentStyle.body.repeat(maxLineLength)}${commentStyle.end}`;
+      this.comments[fileType] = comment;
+    });
+  }
+
+  _getSourceFolders() {
+    const path = this.destinationPath().split('/');
+    return ['src', path[path.length - 1]];
+  }
+
+  _abort(msg) {
+    console.log('ABORTING');
+    return Promise.reject(msg ? msg : 'Step Failed: Aborting');
+  }
+
+  _addComments(path, fileExtension) {
+    glob(`**/*.${fileExtension}`, {cwd: path}, (err, files) => {
+      if (err) {
+        throw err;
+      }
+
+      if (files.length === 0) {
+        return;
+      }
+
+      files.forEach((file) => {
+        const filePath = path + '/' + file;
+        let fileContents = this.fs.read(filePath);
+
+        // TODO: override any comment if the file starts with one (e.g. when the licence changes)
+        // whenever a file starts with our licence header skip for now
+        if (fileContents.startsWith(this.comments[fileExtension])) {
+          return;
+        }
+
+        fileContents = this._findAndRemoveHeader(fileContents, fileExtension);
+
+        const newContents = this.comments[fileExtension] + '\n\n' + fileContents;
```

Comment:
not sure regarding the new line characters. it may has to be on windows `\r\n` since git should check it out with the local ones